Objects/object.c

```diff
@@ -1797,7 +1797,11 @@ PyObject_GenericSetDict(PyObject *obj, PyObject *value, void *context)
                      "not a '%.200s'", Py_TYPE(value)->tp_name);
         return -1;
     }
+#ifdef Py_GIL_DISABLED
+    Py_XDECREF(_Py_atomic_exchange_ptr(dictptr, Py_NewRef(value)));
```

Comment:
I think we need to be consistent about either using atomic exchange/cas or using the object locks to protect the dictptr. I don't think we can safely mix them.